### PR TITLE
Write wide strings to stdout; cleanup logging providers.

### PR DESCRIPTION
**Pull request by jkotalik**
Need #1331 before finalizing tests/PR.

### CODE REVIEW

File: src/AspNetCoreModuleV2/CommonLib/PipeOutputManager.cpp [hunk 1]
```diff
@@ -152,10 +151,14 @@ HRESULT PipeOutputManager::Stop()
 
     // If we captured any output, relog it to the original stdout
     // Useful for the IIS Express scenario as it is running with stdout and stderr
-    if (GetStdOutContent(&straStdOutput))
+    int nChars = MultiByteToWideChar(GetConsoleOutputCP(), 0, m_pipeContents, static_cast<int>(m_numBytesReadTotal), NULL, 0);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/PipeOutputManager.cpp]: Check return value.

File: src/AspNetCoreModuleV2/CommonLib/StringHelpers.cpp [hunk 2]
```diff
@@ -13,3 +14,22 @@ bool ends_with(const std::wstring &source, const std::wstring &suffix, bool igno
     const auto offset = source.length() - suffix.length();
     return CSTR_EQUAL == CompareStringOrdinal(source.c_str() + offset, static_cast<int>(suffix.length()), suffix.c_str(), static_cast<int>(suffix.length()), ignoreCase);
 }
+
+HRESULT to_wide_string(const std::string &source, std::wstring &destination)
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/StringHelpers.cpp]: Can we throw?

Comment (jkotalik) [reply src/AspNetCoreModuleV2/CommonLib/StringHelpers.cpp]: Yeah, I'll refactor to handle throwing.

File: src/AspNetCoreModuleV2/CommonLib/StringHelpers.cpp [hunk 3]
```diff
@@ -18,3 +19,30 @@ bool equals_ignore_case(const std::wstring& s1, const std::wstring& s2)
...
+    {
+        THROW_LAST_ERROR();
+    }
+
+    destination.resize(nChars);
+
+    MultiByteToWideChar(GetConsoleOutputCP(), 0, source.data(), static_cast<int>(source.length()), destination.data(), nChars);
+    if (nChars == 0)
+    {
+        THROW_LAST_ERROR();
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/StringHelpers.cpp]: THROW_LAST_ERROR_IF

File: src/AspNetCoreModuleV2/CommonLib/StringHelpers.cpp [hunk 4]
```diff
@@ -18,3 +19,25 @@ bool equals_ignore_case(const std::wstring& s1, const std::wstring& s2)
...
+    {
+        return L"";
+    }
+
+    std::wstring destination;
+
+    int nChars = MultiByteToWideChar(GetConsoleOutputCP(), 0, source.data(), static_cast<int>(source.length()), NULL, 0);
+    THROW_LAST_ERROR_IF(nChars == 0);
+
+    destination.resize(nChars);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/StringHelpers.cpp]: nChars include '\0' do you need to resize it to nChars -1 in the end ?

File: test/WebSites/StartupExceptionWebSite/Program.cs [hunk 5]
```diff
@@ -61,6 +60,10 @@ public static int Main(string[] args)
                     Console.WriteLine($"Is Console redirection: {Console.IsOutputRedirected}");
                     Console.BackgroundColor = ConsoleColor.Blue;
                     break;
+                case "CheckUTF8":
```
Comment (pakrym) [inline test/WebSites/StartupExceptionWebSite/Program.cs]: Merge this with CheckConsoleFunctions

File: test/Common.FunctionalTests/Inprocess/LoggingTests.cs [hunk 6]
```diff
@@ -200,6 +206,50 @@ public async Task StartupMessagesLogFileSwitchedWhenLogFilePresentInWebConfig()
...
+
+            var response = await deploymentResult.HttpClient.GetAsync(path);
+
+            Assert.Equal(HttpStatusCode.InternalServerError, response.StatusCode);
+
+            StopServer();
+
+            var contents = File.ReadAllText(Helpers.GetExpectedLogName(deploymentResult, _logFolderPath));
+
+            Assert.NotNull(contents);
```
Comment (pakrym) [inline test/Common.FunctionalTests/Inprocess/LoggingTests.cs]: ReadAllText would never return null

File: src/AspNetCoreModuleV2/CommonLib/StringHelpers.cpp [hunk 7]
```diff
@@ -18,3 +19,25 @@ bool equals_ignore_case(const std::wstring& s1, const std::wstring& s2)
...
+    // MultiByteToWideChar returns 0 on failure, which is also the same return value
+    // for empty strings. Preemptive return. 
+    if (source.length() == 0)
+    {
+        return L"";
+    }
+
+    std::wstring destination;
+
+    int nChars = MultiByteToWideChar(GetConsoleOutputCP(), 0, source.data(), static_cast<int>(source.length()), NULL, 0);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/StringHelpers.cpp]: GetConsoleOutputCP() doesn't belong here, take it as an argument

File: src/AspNetCoreModuleV2/CommonLib/BaseOutputManager.h [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        stdoutWrapper(nullptr),
+        stderrWrapper(nullptr),
+        m_enableNativeRedirection(enableNativeLogging)
+    {
+        InitializeSRWLock(&m_srwLock);
+    }
+    ~BaseOutputManager() {}
+
+    void
+        StartRedirectionWithExceptionMessage(
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/BaseOutputManager.h]: nit: formatting of this file

File: src/AspNetCoreModuleV2/CommonLib/BaseOutputManager.h [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            std::wstring exceptionMessage)
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/BaseOutputManager.h]: having the error message as an argument is a bit ugly

File: src/AspNetCoreModuleV2/CommonLib/BaseOutputManager.h [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    std::unique_ptr<StdWrapper> stdoutWrapper;
+    std::unique_ptr<StdWrapper> stderrWrapper;
+
+    void
+        TryOperation(void(*func)(BaseOutputManager* outputManager),
+            std::wstring exceptionMessage)
+    {
+        try
+        {
+            func(this);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/BaseOutputManager.h]: You can capture `this` in lambda.

File: src/AspNetCoreModuleV2/CommonLib/FileOutputManager.cpp [hunk 11]
```diff
@@ -57,68 +47,58 @@ FileOutputManager::Start()
         // ERROR_ACCESS_DENIED means there is a console already present.
         if (GetLastError() != ERROR_ACCESS_DENIED)
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/FileOutputManager.cpp]: THROW_LAST_ERROR_IO(GetLastError() != ERROR_ACCESS_DENIED);

File: src/AspNetCoreModuleV2/CommonLib/BaseOutputManager.h [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        }
+        catch (ResultException& exception)
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/BaseOutputManager.h]: This wouldn't catch non `ResultException` like std::runtime_error thrown by `create_directories`

File: src/AspNetCoreModuleV2/CommonLib/BaseOutputManager.h (hunk 10 above)
Comment (jkotalik) [reply src/AspNetCoreModuleV2/CommonLib/BaseOutputManager.h]: not sure how to work around: https://stackoverflow.com/questions/31611415/converting-a-lambda-expression-with-variable-capture-to-a-function-pointer

Comment (pakrym) [reply src/AspNetCoreModuleV2/CommonLib/BaseOutputManager.h]: Did you try using templated function?

```
template<typename Functor>
void f(Func ff)
{
ff()?
}
```

File: src/AspNetCoreModuleV2/CommonLib/StringHelpers.cpp [hunk 13]
```diff
@@ -18,3 +19,25 @@ bool equals_ignore_case(const std::wstring& s1, const std::wstring& s2)
...
+
+    std::wstring destination;
+
+    int nChars = MultiByteToWideChar(codePage, 0, source.data(), static_cast<int>(source.length()), NULL, 0);
+    THROW_LAST_ERROR_IF(nChars == 0);
+
+    destination.resize(nChars);
+
+    MultiByteToWideChar(codePage, 0, source.data(), static_cast<int>(source.length()), destination.data(), nChars);
+    THROW_LAST_ERROR_IF(nChars == 0);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/StringHelpers.cpp]: You are not checking the right thing here.

File: src/AspNetCoreModuleV2/AspNetCore/dllmain.cpp [hunk 14]
```diff
@@ -101,7 +101,7 @@ HRESULT
     {
         g_hEventLog = RegisterEventSource(nullptr, ASPNETCORE_EVENT_PROVIDER);
     }
-
+
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/AspNetCore/dllmain.cpp]: revert

File: src/AspNetCoreModuleV2/CommonLib/FileOutputManager.h [hunk 15]
```diff
@@ -4,38 +4,27 @@
...
 
-class FileOutputManager : public IOutputManager
+class FileOutputManager : public BaseOutputManager
 {
-    #define FILE_FLUSH_TIMEOUT 3000
     #define MAX_FILE_READ_SIZE 30000
 public:
-    FileOutputManager();
-    FileOutputManager(bool fEnableNativeLogging);
+    FileOutputManager(PCWSTR pwzApplicationPath, PCWSTR pwzStdOutLogFileName);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/FileOutputManager.h]: Any reason not to take strings?

File: src/AspNetCoreModuleV2/CommonLib/exceptions.h [hunk 16]
```diff
@@ -59,6 +59,7 @@
 #define FINISHED_LAST_ERROR_IF(condition)                       do { if (condition) { hr = LogLastError(LOCATION_INFO); _GOTO_FINISHED(); }} while (0, 0)
 #define FINISHED_LAST_ERROR_IF_NULL(ptr)                        do { if ((ptr) == nullptr) { hr = LogLastError(LOCATION_INFO); _GOTO_FINISHED(); }} while (0, 0)
 
+#define THROW(hr)                                               do { _HR_RET(hr); ThrowResultException(LOCATION_INFO, LogHResultFailed(LOCATION_INFO, __hrRet)); } while (0, 0)
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/exceptions.h]: THROW_HR

File: test/Common.FunctionalTests/Inprocess/LoggingTests.cs [hunk 17]
```diff
@@ -222,6 +227,48 @@ public async Task OutOfProcessReadsLogConfiguration()
...
+
+            Assert.Equal(HttpStatusCode.InternalServerError, response.StatusCode);
+
+            StopServer();
+
+            var contents = File.ReadAllText(Helpers.GetExpectedLogName(deploymentResult, _logFolderPath));
+
+            Assert.Contains("彡⾔", contents);
+
+            EventLogHelpers.VerifyEventLogEvent(deploymentResult, "彡⾔");
```
Comment (pakrym) [inline test/Common.FunctionalTests/Inprocess/LoggingTests.cs]: Verify which event stdout is part of

**Pull request merged**